src/wasm-type.h

```diff
@@ -48,21 +49,52 @@ class Type {
   // But converting raw uint32_t is more dangerous, so make it explicit
   constexpr explicit Type(uint32_t id) : id(id){};
 
+  // Construct from lists of elementary types
+  Type(std::initializer_list<Type> types);
+  explicit Type(const std::vector<Type>& types);
+
+  // Accessors
+  size_t size() const;
+  const std::vector<Type>& expand() const;
+
+  // Predicates
+  bool isSingle() const { return id >= i32 && id < _last_value_type; }
+  bool isMulti() const { return id >= _last_value_type; }
+  bool isConcrete() const { return id >= i32; }
+  bool isInteger() const { return id == i32 || id == i64; }
+  bool isFloat() const { return id == f32 || id == f64; }
+  bool isVector() const { return id == v128; };
+  bool isRef() const { return id == anyref || id == exnref; }
+
   // (In)equality must be defined for both Type and ValueType because it is
   // otherwise ambiguous whether to convert both this and other to int or
   // convert other to Type.
-  bool operator==(const Type& other) { return id == other.id; }
+  bool operator==(const Type& other) const { return id == other.id; }
+  bool operator==(const ValueType& other) const { return id == other; }
+  bool operator!=(const Type& other) const { return id != other.id; }
+  bool operator!=(const ValueType& other) const { return id != other; }
 
-  bool operator==(const ValueType& other) { return id == other; }
-
-  bool operator!=(const Type& other) { return id != other.id; }
+  // Allows for using Types in switch statements
+  constexpr operator uint32_t() const { return id; }
+  std::string toString() const;
+};
 
-  bool operator!=(const ValueType& other) { return id != other; }
+struct ParamType {
```

Comment:
Are ParamType and ResultType just for printing purposes? If so maybe add a comment. Or is there another use?

Reply:
Yes, I'll add comments.